src/Consumer/AbstractConsumer.php

```diff
@@ -14,70 +15,87 @@
 
 abstract class AbstractConsumer implements ConsumerInterface
 {
-    protected $container;
+    use LoggerAwareTrait;
 
-    public function __construct(ContainerInterface $container)
+    protected $logger;
+    protected $validator;
+    protected $registry;
+
+    public function __construct(ValidatorInterface $validator, Registry $registry)
     {
-        $this->container = $container;
+        $this->validator = $validator;
+        $this->registry = $registry;
     }
 
     public function execute(AMQPMessage $message)
     {
-        $logger = $this->getLogger();
-
         try {
             $data = \GuzzleHttp\json_decode($message->body, true);
         } catch (\Exception $e) {
-            $logger->error('Message is not valid JSON', [
+            $this->getLogger()->error('Message is not valid JSON', [
                 'message' => $message->body,
             ]);
 
-            return true;
+            return ConsumerInterface::MSG_ACK;
         }
 
-        $violations = $this->getValidator()->validate($data, new Assert\Collection([
-            'allowExtraFields' => false,
-            'allowMissingFields' => false,
-            'fields' => $this->configureDataConstraints(),
-        ]));
-
-        if ($violations->count() > 0) {
-            $messages = [];
-
-            /** @var ConstraintViolationInterface $violation */
-            foreach ($violations as $violation) {
-                $messages[$violation->getPropertyPath()][] = $violation->getMessage();
-            }
-
-            $logger->error('Message structure is not valid', [
+        if ($messages = $this->validate($data)) {
+            $this->getLogger()->error('Message structure is not valid', [
                 'message' => $message->body,
                 'violations' => $messages,
             ]);
 
-            return true;
+            return ConsumerInterface::MSG_ACK;
         }
 
         return $this->doExecute($data);
     }
 
     public function writeln($name, $message): void
     {
-        echo $name.' | '.$message."\n";
+        echo $name.' | '.$message.PHP_EOL;
     }
 
-    protected function getLogger(): LoggerInterface
+    public function getLogger(): LoggerInterface
```

Comment:
Does it really need to be public?

Reply:
By being public it is easier to test the "if" by verifying the call to the methods of the logger because the function "execute" always returns a ConsumerInterface :: MSG_ACK. I can not trust the return of the function to check that the execution of the code is correct